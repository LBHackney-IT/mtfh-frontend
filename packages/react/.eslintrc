{
  "extends": "@lbh-mtfh/eslint-config",
  "parserOptions": {
    "project": "tsconfig.json"
  },
  "rules": {
    "prefer-arrow-callback": "off",
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": ["**/*/rollup.config.js"]
      }
    ],
    "jsx-a11y/anchor-is-valid": [
      "error",
      {
        "components": ["Link"],
        "aspects": ["invalidHref"]
      }
    ]
  }
}
