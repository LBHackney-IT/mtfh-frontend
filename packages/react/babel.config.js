const babel = require("../../babel.config")

module.exports = babel
