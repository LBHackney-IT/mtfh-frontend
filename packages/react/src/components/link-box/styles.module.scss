.link-overlay a {
  position: static;

  &:before {
    content: "";
    cursor: inherit;
    display: block;
    position: absolute;
    top: 0;
    left: 0;
    z-index: 0;
    width: 100%;
    height: 100%;
  }
}

.link-box {
  position: relative;

  *:not(.link-overlay) > a[href] {
    position: relative;
    z-index: 1;
  }
}
