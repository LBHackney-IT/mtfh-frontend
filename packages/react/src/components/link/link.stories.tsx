import React from "react";

import { Text } from "../text";
import { Link } from "./link";

export default {
  title: "Link",
  component: Link,
};

export const Base = () => {
  return <Link href="/">Home</Link>;
};

export const DangerLink = () => {
  return (
    <Link href="/delete" variant="danger">
      Delete Record
    </Link>
  );
};

export const MutedLink = () => {
  return (
    <Link href="/activities" variant="muted">
      Past activities
    </Link>
  );
};

export const TextColorLink = () => {
  return (
    <Text>
      Read more about{" "}
      <Link href="/mark" variant="text-colour">
        Mark
      </Link>{" "}
      in our new{" "}
      <Link href="/post/1" variant="text-colour">
        blog post
      </Link>
      .
    </Text>
  );
};

export const NoUnderlineLink = () => {
  return (
    <Link href="/" variant="no-underline">
      Home
    </Link>
  );
};

export const BackLink = () => {
  return (
    <Link href="/back" variant="back-link">
      Home
    </Link>
  );
};

export const ExternalLink = () => {
  return (
    <Link href="http://google.com" isExternal>
      Google
    </Link>
  );
};
