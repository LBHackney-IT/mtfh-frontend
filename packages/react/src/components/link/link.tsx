import React, { forwardRef } from "react";
import type * as Polymorphic from "@radix-ui/react-polymorphic";
import cn from "classnames";

import { widthOverrides } from "../../utils";
import styles from "./styles.module.scss";

export interface LinkProps {
  variant?:
    | "link"
    | "danger"
    | "text-colour"
    | "muted"
    | "back-link"
    | "no-underline"
    | "native";
  isExternal?: boolean;
  override?: number;
}

export type LinkComponent = Polymorphic.ForwardRefComponent<"a", LinkProps>;

export const Link: LinkComponent = forwardRef(function Link(
  {
    as: LinkComp = "a",
    variant = "link",
    isExternal = false,
    className,
    rel,
    target,
    override,
    ...props
  },
  ref,
) {
  const linkClasses = cn(
    variant !== "native" && {
      [`${styles.govukLink} ${styles.lbhLink}`]: variant !== "back-link",
      [`${styles.govukBackLink} ${styles.lbhBackLink}`]: variant === "back-link",
      [styles[`lbh-link--${variant}`]]: variant !== "link" && variant !== "back-link",
      [styles.lbhLinkNoVisitedState]:
        !isExternal && !["danger", "muted", "text-colour"].includes(variant),
    },
    widthOverrides(override),
    className,
  );

  return (
    <LinkComp
      ref={ref}
      className={linkClasses}
      rel={isExternal ? "noopener noreferrer" : rel}
      target={isExternal ? "_blank" : target}
      {...props}
    />
  );
});
