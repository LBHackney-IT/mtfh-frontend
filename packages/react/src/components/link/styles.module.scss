@import "~lbh-frontend/lbh/base";
@import "~govuk-frontend/govuk/core/links";
@import "~lbh-frontend/lbh/core/links";
@import "~lbh-frontend/lbh/components/lbh-back-link/back-link";

.lbh-link--no-underline {
  composes: govuk-link--no-underline;
}
