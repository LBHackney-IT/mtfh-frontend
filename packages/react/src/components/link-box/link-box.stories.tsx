import React from "react";

import { Center } from "../center";
import { Link } from "../link";
import { LinkBox, LinkOverlay } from "./link-box";

export default {
  title: "LinkBox",
  component: LinkBox,
};

export const Base = () => {
  return (
    <LinkBox
      style={{
        border: "2px solid grey",
        borderRadius: "6px",
        width: "200px",
        height: "200px",
      }}
    >
      <Center style={{ height: "100%" }}>
        <LinkOverlay>
          <Link href="/">Link</Link>
        </LinkOverlay>
      </Center>
    </LinkBox>
  );
};
