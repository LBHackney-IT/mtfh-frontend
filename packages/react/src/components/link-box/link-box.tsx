import React, { ReactElement, forwardRef } from "react";
import type * as Polymorphic from "@radix-ui/react-polymorphic";
import cn from "classnames";

import { widthOverrides } from "../../utils";
import { LinkProps } from "../link";
import styles from "./styles.module.scss";

export interface LinkOverlayProps {
  children: ReactElement<LinkProps>;
  override?: number;
}

export type LinkOverlayComponent = Polymorphic.ForwardRefComponent<
  "div",
  LinkOverlayProps
>;

export const LinkOverlay: LinkOverlayComponent = forwardRef(function LinkOverlay(
  { as: LinkOverlayComp = "div", className, override, ...props },
  ref,
) {
  return (
    <LinkOverlayComp
      ref={ref}
      className={cn(styles.linkOverlay, widthOverrides(override), className)}
      {...props}
    />
  );
});

export interface LinkBoxProps {
  override?: number;
}

export type LinkBoxComponent = Polymorphic.ForwardRefComponent<"div", LinkBoxProps>;

export const LinkBox: LinkBoxComponent = forwardRef(function LinkBox(
  { as: LinkBoxComp = "div", className, override, ...props },
  ref,
) {
  return (
    <LinkBoxComp
      ref={ref}
      className={cn(styles.linkBox, widthOverrides(override), className)}
      {...props}
    />
  );
});
