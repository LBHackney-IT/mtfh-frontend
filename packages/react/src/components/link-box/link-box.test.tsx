import React from "react"
import { testA11y } from "@hackney/mtfh-test-utils"
import { render } from "@testing-library/react"

import { Link } from "../link"
import { LinkBox, LinkOverlay } from "./link-box"

test("it renders correctly", async () => {
  const { container } = render(
    <LinkBox>
      <LinkOverlay>
        <Link href="https://localhost">Link</Link>
      </LinkOverlay>
    </LinkBox>,
  )

  expect(container).toMatchSnapshot()
  await testA11y(container)
})
