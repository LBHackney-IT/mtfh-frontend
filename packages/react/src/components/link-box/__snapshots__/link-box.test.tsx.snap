// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders correctly 1`] = `
<div>
  <div
    class="linkBox"
  >
    <div
      class="linkOverlay"
    >
      <a
        class="govukLink lbhLink lbhLinkNoVisitedState"
        href="https://localhost"
      >
        Link
      </a>
    </div>
  </div>
</div>
`;
